### PR TITLE
Feature/mesoscale mpmd fix

### PR DESCRIPTION
**Pull request by PerryShafran-NOAA**
<b>Note to developers: You must use this PR template!</b>

## Description of Changes

> Please include a summary of the changes and the related GitHub issue(s). Please also include relevant motivation and context.

This PR makes the mpmd fix regarding Bugzilla 1547 for the mesoscale deterministic (NAM and RAP) component.  Components fixed were the grid2obs stats, and all of the plots.  The precip and snowfall stats are being worked out, but I am submitting what works so we can get those changes in.  

## Developer Questions and Checklist
* Is this a high priorty PR? If so, why and is there a date it needs to be merged by?

Try to get it merged by COB 12/13 in order to complete before WCOSS2 machine outage.  
* Do you have any planned upcoming annual leave/PTO?

No, except potentially Jan 2-3.

* Are there any changes needed for when the jobs are supposed to run?

Not for now.  We'll work up these changes for a future PR.  
  
- [ ] The code changes follow [NCO's EE2 Standards](https://www.nco.ncep.noaa.gov/idsb/implementation_standards/ImplementationStandards.v11.0.0.pdf).
- [ ] Developer's name is removed throughout the code and have used `${USER}` where necessary throughout the code.
- [ ] References the feature branch for `HOMEevs` are removed from the code.
- [ ] J-Job environment variables, COMIN and COMOUT directories, and output follow what has been [defined](https://docs.google.com/document/d/1JWg_4q80aYmmAoD21GFjp9R9y5-3w7WGM3-0HJk0Pjs/edit#heading=h.7ysbr191vzu4) for EVS.
- [ ] Jobs over 15 minutes in runtime have restart capability.
- [ ] If applicable, changes in the `dev/drivers/scripts` or `dev/modulefiles` have been made in the corresponding `ecf/scripts` and `ecf/defs/evs-nco.def`? 
- [ ] Jobs contain the approriate file checking and don't run METplus for any missing data.
- [ ] Code is using METplus wrappers structure and not calling MET executables directly.
- [ ] Log is free of any ERRORs or WARNINGs.

## Testing Instructions

> Please include testing instructions for the PR assignee. Include all relevant input datasets needed to run the tests.

1) Clone my fork https://github.com/PerryShafran-NOAA/EVS.git.
2) Checkout the feature/mesoscale_mpmd_fix branch.
3) Link the fix directory.
4) Set export EVSINspcotlk=/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v2.0/prep/cam for both the dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats.sh and jevs_mesoscale_rap_grid2obs_stats.sh drivers.
5) Run both the above drivers using qsub -v vhr=07.
6) In dev/drivers/scripts/plots/mesoscale, run the jevs_mesoscale_grid2obs_plots.sh and jevs_mesoscale_headline_plots.sh drivers using qsub -v vhr=01.
7) In dev/drivers/scripts/plots/mesoscale, run the jevs_mesoscale_precip_plots.sh and jevs_mesoscale_snowfall_plots.sh drivers using qsub -v vhr=13.

### CODE REVIEW

File: scripts/stats/mesoscale/exevs_mesoscale_nam_grid2obs_stats.sh [hunk 1]
```diff
@@ -316,6 +346,10 @@ echo "*****************************"
      python $USHevs/mesoscale/mesoscale_stats_grid2obs_create_poe_job_scripts.py
      export err=$?; err_chk
   fi
+ 
+ Create Gather 3 Working Directories
```
Comment (malloryprow) [inline scripts/stats/mesoscale/exevs_mesoscale_nam_grid2obs_stats.sh]: Add `#` at start of line

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Two things I forgot to add regarding the instructions above:

1) Set COMIN for all jobs to the emc.vpppg directory.
2) I recommend setting off plots early because they take so long.  You do not need stats to finish to run the plots.

File: scripts/stats/mesoscale/exevs_mesoscale_nam_grid2obs_stats.sh (hunk 1 above)
Comment (PerryShafran-NOAA) [reply scripts/stats/mesoscale/exevs_mesoscale_nam_grid2obs_stats.sh]: Done.

**Comment by malloryprow** [comment]
COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/evs/v2.0/stats/mesoscale

### 1. jevs_mesoscale_nam_grid2obs_stats (vhr=07)
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_nam_grid2obs_stats_00.o209081464
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_nam_grid2obs_stats_00.209081464.dbqs01

### 2. jevs_mesoscale_rap_grid2obs_stats (vhr=07)
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/stats/mesoscale/jevs_mesoscale_rap_grid2obs_stats_00.o209081468
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_rap_grid2obs_stats_00.209081468.dbqs01

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Can you add the COMOUT directory?  We'll need to see the final files to figure out whether their filesize is good.  

Thanks!

**Comment by malloryprow** [comment]
Derp! Yup added to original comment.

**Comment by PerryShafran-NOAA** [comment]
Thus far, I am seeing the subdirectories being used to process data for both the nam and rap stats.  So that's a good start!

**Comment by malloryprow** [comment]
I'm seeing lots of errors in the logs.

> ERROR  : 
> ERROR  : PointStatVxOpt::set_vx_pd() -> At least one output masking region must be requested in "mask.grid", "mask.poly", "mask.sid", or "mask.llpnt".
> ERROR  : 
> ERROR  : 
> ERROR  : PointStatVxOpt::set_vx_pd() -> At least one output masking region must be requested in "mask.grid", "mask.poly", "mask.sid", or "mask.llpnt".
> ERROR  : 

is what I'm seeing a lot.

**Comment by PerryShafran-NOAA** [comment]
Darn it.  What'd I forget???  Let me check this out.

**Comment by PerryShafran-NOAA** [comment]
@malloryprow OK, here's an explanation.

You get this error when spc_otlk files are missing.  And I am seeing that in emc.vpppg/evs/v2.0/prep/cam/spc_otlk.20241211, day 2 files are missing, and in spc_otlk.20241212, day 1 files are missing.  I did not see any of these errors in yesterday's test in my directory, but I have them in today's, just like yours.  

I have no idea why file checks aren't here for those files, but I'm tagging @MarcelCaron-NOAA in case he can provide any insight.  He may have similar file checks in cam.  

If possible, since this isn't quite in the scope of this PR, we can add this to the Fixes and Additions and I'll see where I can add file checks.  @AliciaBentley-NOAA @AndrewBenjamin-NOAA

**Comment by PerryShafran-NOAA** [comment]
FWIW, the errors appear in the emc.vpppg run as well.

See /lfs/h2/emc/ptmp/emc.vpppg/output/jevs_mesoscale_nam_grid2obs_stats_00.o209056259

**Comment by malloryprow** [comment]
> Thus far, I am seeing the subdirectories being used to process data for both the nam and rap stats. So that's a good start!

Yup, I'm seeing that too!

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Your stats output matches my files for both NAM and RAP grid2obs.  They also match what is in emc.vpppg.  

I would be OK with calling stats validated for now, and moving on to the plot jobs.  Please submit them when you can; they are long plot jobs (esp. precip and snowfall).

**Comment by malloryprow** [comment]
### FULL TEST

COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/evs/v2.0/plots/mesoscale

### 1. jevs_mesoscale_grid2obs_plots
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_grid2obs_plots.o209089950
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_grid2obs_plots.209089950.dbqs01

### 2. jevs_mesoscale_headline_plots
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_headline_plots.o209090107
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_headline_plots.209090107.dbqs01

### 3. jevs_mesoscale_precip_plots
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_precip_plots.o209090157
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_precip_plots.209090157.dbqs01

### 4. jevs_mesoscale_snowfall_plots
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_snowfall_plots.o209090200
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_snowfall_plots.209090200.dbqs01

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Subdirectories are properly being used for all 4 plot jobs.

**Comment by malloryprow** [comment]
Can you increase the memory in jevs_mesoscale_headline_plots (dev and ecf)?

**Comment by PerryShafran-NOAA** [comment]
I adjusted the memory to 100 MB.  I don't know if that's what you wanted, but it's more.

I also adjusted the time so that the two have the same time, and I am using 30 minutes.  They were different before.

**Comment by PerryShafran-NOAA** [comment]
@malloryprow The mesoscale grid2obs plots job is done.  The tarball is the correct size.  No errors or warnings.

**Comment by malloryprow** [comment]
NCO recommends for jobs running 3-10 minutes at 15 min walltime.

100GB is a bit too much the re-run used 25798128kb/25GB, so I'd say like 35GB is fine.

Also can remove `exclhost` we aren't requesting 500GB memory or 128 cpus.

Re-run info

### 1. jevs_mesoscale_headline_plots
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_headline_plots.o209100444
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_headline_plots.209100444.dbqs01
**COMOUT:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/evs/v2.0/plots/mesoscale

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Thank you! I made adjustments based on those recommendations.

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Everything looks good with the headline plots.

**Comment by AndrewBenjamin-NOAA** [comment]
> @malloryprow OK, here's an explanation.
> 
> You get this error when spc_otlk files are missing. And I am seeing that in emc.vpppg/evs/v2.0/prep/cam/spc_otlk.20241211, day 2 files are missing, and in spc_otlk.20241212, day 1 files are missing. I did not see any of these errors in yesterday's test in my directory, but I have them in today's, just like yours.
> 
> I have no idea why file checks aren't here for those files, but I'm tagging @MarcelCaron-NOAA in case he can provide any insight. He may have similar file checks in cam.
> 
> If possible, since this isn't quite in the scope of this PR, we can add this to the Fixes and Additions and I'll see where I can add file checks. @AliciaBentley-NOAA @AndrewBenjamin-NOAA

@PerryShafran-NOAA I am fine with leaving this for a follow up PR.

I did a quick look in exevs_mesoscale_rap_grid2obs_stats.sh and exevs_mesoscale_anm_grid2obs_stats.sh, and I think the disconnect might be that spc_otlk is defined in these scripts(in the NEST_LIST for loop). No logic for spc_otlk is present to mesoscale_check_input_data.py, so no checking is actually taking place. That may not be exactly what's going on, but given this error comes up specfically when scp_otlk files are missing, it would be a good place to start in a future PR.

**Comment by PerryShafran-NOAA** [comment]
@malloryprow The precip plot tarball is good.  No errors or warnings in the log.

**Comment by PerryShafran-NOAA** [comment]
@malloryprow Finally, the snowfall plot job is finished.  Tarball is the right size, no warnings or errors.

I think we have validated everything!

**Comment by malloryprow** [comment]
Great! I think I'd like to test restart for the plotting jobs; I know there is a Bugzilla for the grid-to-obs stats. The plots jobs are quite long and I want to make sure that they still work with the MPMD bugzilla fix.

**Comment by PerryShafran-NOAA** [comment]
@malloryprow OK, that should be fine.

**Comment by PerryShafran-NOAA** [comment]
It should also be noted that I plan to mimic the cam verification and split out plot jobs into 31days and 90days plots.  I don't know how that's working out in cam, but that should make plot jobs shorter.

**Comment by malloryprow** [comment]
### RESART TEST

COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/evs/v2.0/plots/mesoscale. The headline plot runs for about ~6 minutes, so no need for restart testing. 

### 1. jevs_mesoscale_grid2obs_plots
**Interrupted Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_grid2obs_plots.o209307816
**Interrupted DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_grid2obs_plots.209307816.dbqs01
**Restart Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_grid2obs_plots.o209309024
**Restart DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_grid2obs_plots.209309024.dbqs01

### 2. jevs_mesoscale_precip_plots
**Interrupted Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_precip_plots.o209307902
**Interrupted DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_precip_plots.209307902.dbqs01
**Restart Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_precip_plots.o209311376
**Restart DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_precip_plots.209311376.dbqs01

### 3. jevs_mesoscale_snowfall_plots
**Interrupted Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_snowfall_plots.o209307915
**Interrupted DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_snowfall_plots.209307915.dbqs01
**Restart Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr619/EVS/dev/drivers/scripts/plots/mesoscale/jevs_mesoscale_snowfall_plots.o209311377
**Restart DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_mesoscale_snowfall_plots.209311377.dbqs01

**Comment by PerryShafran-NOAA** [comment]
@malloryprow I do not know how long you ran the interrupted job for.  However, the mesoscale_create_child_workdirs.py process seems to take a long time (as does the job script creation).  Thus, the interrupted job did not get to do any plot generation. Thus, you'll have to run the interrupted job a bit longer if you want to get some plots; otherwise, it's not true restart since no plots were done.

**Comment by malloryprow** [comment]
The grid2obs plots ran for 25 minutes and the other two for an hour.

**Comment by malloryprow** [comment]
I guess it is okay to not test restart for the plots either if they are getting split up. We can test restart then. The jobs running as long as they did, but not getting to any plots yet is a little concerning, especially for the jobs that had run for an hour. I think the job being split into the 31 days and 90 days time period should help that.

**Comment by PerryShafran-NOAA** [comment]
OK, precip was OK, there were plots in the restart directory.  For grid2obs and snowfall they were not.  

A lot of time is taken to both create the job scripts and also to create the child workdirs.  I copied those processes from Marcel (like most other things in this component).  I don't know for sure why they take so long, but they do.  I haven't been closely following your experiences in cam, but Marcel did this split - did that significantly shorten those plot job?  

Splitting the plot jobs would definitely have fewer job scripts and child directories to process, but a lot of the time is spent in these two processes.  Restart jobs would have to re-start these processes (since they happen in the $DATA directory), and thus restart time wouldn't be any shorter than the full run time.  

I had originally thought that the 31/90 day split was in the "nice to have" category, but given this issue it really should be in the "must have" category to shorten these processes, I think.

**Comment by malloryprow** [comment]
Yeah, hopefully it won't be too bad do this. I did add it to the Fixes and Additions document.

**Review by malloryprow — APPROVED** [review]
Changes are good and testing successful.

Thanks @PerryShafran-NOAA!

The one item for increasing the memory is on the Fixes and Additions document, so please check it off.

**Comment by MarcelCaron-NOAA** [comment]
> OK, precip was OK, there were plots in the restart directory. For grid2obs and snowfall they were not.
> 
> A lot of time is taken to both create the job scripts and also to create the child workdirs. I copied those processes from Marcel (like most other things in this component). I don't know for sure why they take so long, but they do. I haven't been closely following your experiences in cam, but Marcel did this split - did that significantly shorten those plot job?
> 
> Splitting the plot jobs would definitely have fewer job scripts and child directories to process, but a lot of the time is spent in these two processes. Restart jobs would have to re-start these processes (since they happen in the $DATA directory), and thus restart time wouldn't be any shorter than the full run time.
> 
> I had originally thought that the 31/90 day split was in the "nice to have" category, but given this issue it really should be in the "must have" category to shorten these processes, I think.

@PerryShafran-NOAA Yup the split made a big difference for cam.  I think cutting the jobs in half cuts many of the longer processes (writing cards, working directories, cat'ing output ...) in half.

**Comment by malloryprow** [comment]
BTW I killed the jobs.

**Review by AndrewBenjamin-NOAA — APPROVED** [review]
I reviewed the changes and approve this PR. I agree with the conversation around splitting up the plotting jobs is a great idea given Marcel's comments.

**Pull request merged**